### PR TITLE
Fix #1298: Introduce app auto expiration

### LINKED ISSUE #1298: Alpha app should expire after fixed amount of time
Since the alpha version of the app is a pre-release, we should "bake in" an automatic expiration timer. I think 3 months should be more than sufficient. Users who use a version of the alpha app for more than three months will be prevented from continuing using it until they upgrade via the Play Store.

The production version of the app will not contain this limitation, and will have a more controlled upgrade policy built into it.

### PR DESCRIPTION
**Pull request by BenHenning**
Fixes #1298.
Fix part of #973.

Introduce support for automatically expiring the app after a certain date has been exceeded.

This is meant to be a stop-gap solution for pre-production releases of the app until we have proper app deprecation support & a regular release process.

We configure the expiration date using an AndroidManifest meta-data value that we will set in preparation for each individual Android launch, and an overall boolean to turn the check on/off. The implementation includes a refactoring of the onboarding flow to be a bit more general (from the app module's perspective, the onboarding flow can include an app deprecation case which occurs when the current time exceeds the manifest time). The actual on-disk file for storing whether the onboarding flow has been completed by the user is staying the same for backward compatibility.

The mechanism for retrieving metadata has been abstracted to allow in-test fake replacement to make testing on Espresso/Robolectric easier for app module tests like SplashActivityTest. domain module tests that run only on Robolectric can use the production module + shadows. Since the fake is trivial & the real implementation is covered in ``AppStartupStateControllerTest``, no additional retriever tests have been added.

Finally, this PR also enables splash tests for CI since they've been stabilized in #1623 & thoroughly updated in this PR. I've confirmed that the changes pass on both Espresso & Robolectric.

Screenshot of the new UI:
image

### CODE REVIEW

File: domain/src/test/java/org/oppia/domain/onboarding/AppStartupStateControllerTest.kt [hunk 1]
```diff
@@ -42,16 +49,17 @@ import org.oppia.util.logging.GlobalLogLevel
...
 import javax.inject.Qualifier
 import javax.inject.Singleton
 import kotlin.coroutines.EmptyCoroutineContext
 
-/** Tests for [OnboardingFlowController]. */
+/** Tests for [AppStartupStateController]. */
 @RunWith(AndroidJUnit4::class)
 @Config(manifest = Config.NONE)
-class OnboardingFlowControllerTest {
+class AppStartupStateControllerTest {
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/onboarding/AppStartupStateControllerTest.kt:56]: Tests to add:
- Initial app state, deprecation enabled, not deprecated before deprecation date
- Initial app state, deprecation enabled, deprecated after deprecation date
- Initial app state, deprecation disabled, not deprecated before deprecation date
- App opened before, onboarding flow incomplete, deprecation enabled, not deprecated before deprecation date
- App opened before, onboarding flow incomplete, deprecated after deprecation date
- App opened before, onboarding flow completed, deprecation enabled, not deprecated before deprecation date
- App opened before, onboarding flow completed, deprecated after deprecation date

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -12,6 +12,7 @@
     android:roundIcon="@mipmap/ic_launcher_round"
     android:supportsRtl="true"
     android:theme="@style/OppiaTheme">
+    <meta-data android:name="expiration_date" android:value="2020-09-01" />
```
Comment (BenHenning) [inline app/src/main/AndroidManifest.xml:18]: Functionality to add: disabled flag rather than only relying on the date (so that it can always be disabled for developers by default).

File: app/src/main/java/org/oppia/app/splash/SplashActivityPresenter.kt [hunk 3]
```diff
@@ -7,20 +7,24 @@ import androidx.lifecycle.Observer
...
+import org.oppia.domain.onboarding.AppStartupStateController
 import org.oppia.util.data.AsyncResult
 import org.oppia.util.logging.Logger
 import javax.inject.Inject
 
+private const val AUTO_DEPRECATION_NOTICE_DIALOG_FRAGMENT_TAG = "auto_deprecation_notice_dialog"
+
 /** The presenter for [SplashActivity]. */
 @ActivityScope
 class SplashActivityPresenter @Inject constructor(
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/splash/SplashActivityPresenter.kt:25]: Tests to add:
- Open app, initial open, expiration enabled, before expiration date, intents to onboarding flow
- Open app, initial open, expiration enabled, after expiration date, intents to deprecation dialog
- Open app, initial open, expiration enabled, after expiration date, click on close dialog, ends splash activity
- Open app, initial open, expiration disabled, after expiration date, intents to onboarding flow
- Open app, reopen, expiration enabled, before expiration date, intents to profile chooser
- Open app, reopen, expiration enabled, after expiration date, intents to deprecation dialog

**Comment by BenHenning** [comment]
Reminder to self to update this with the latest develop.

File: app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (aggarwalpulkit596) [reply app/src/main/AndroidManifest.xml:18]: should the value be hardcoded in manifest file? anyone can read these values

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -361,4 +361,21 @@
...
+    Unsupported app version
+  </string>
+  <string
+    name="unsupported_app_version_dialog_message"
+    description="The main message of the dialog shown when a user's pre-release app has expired.">
+    This version of the app is no longer supported. Please update it through the Play Store.
+  </string>
+  <string
+    name="unsupported_app_version_dialog_close_button_text"
+    description="The close button text of the dialog shown when a user's pre-release app has expired.">
```
Comment (rt4914) [inline app/src/main/res/values/strings.xml:378]: Any specific reason behind mentioning this description? Is this related to accessibility?

**Review by rt4914 — APPROVED** [review]
LGTM, the application works correctly under various expiration dates.

**Review by MohamedMedhat1998 — APPROVED** [review]
I have tried changing the date of my emulator and set `automatic_app_expiration_enabled` to `true` in the manifest file and got the expected behavior (the dialogue)  
Just 1 suggestion, you can add the link of the app on the play store in the dialogue, so that we have **Close** and **Update** buttons.  
**Update** button should redirect the user to the play store.  
Other than that, LGTM.

**Review by aggarwalpulkit596 — APPROVED** [review]

**Comment by aggarwalpulkit596** [comment]
@BenHenning i agree with Mohammed we should add a button to update the app although i know this wouldn't be feasible for this particular

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (BenHenning) [reply app/src/main/res/values/strings.xml:378]: It's related to translations, actually. This is something we should be introducing with all strings since it provides translators context when translating the string contents. This will become a priority later in the year, so I'm starting to create some examples for reference for future PRs.

**Comment by BenHenning** [comment]
The update button makes sense, thanks @MohamedMedhat1998. However, we can't actually implement that for pre-release since the listing on the Play Store isn't yet public. This dialog will also be removed before we launch publicly. In the future, we will have other 'update flows' that will include the link.

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (rt4914) [reply app/src/main/res/values/strings.xml:378]: Oh great. This is something new for me. Thanks for explaining.

**Comment by BenHenning** [comment]
Thanks everyone!
